libr/bin/dwarf.c

```diff
@@ -1318,321 +1486,537 @@ static void r_bin_dwarf_dump_debug_info(FILE *f, const RBinDwarfDebugInfo *inf)
-				r_bin_dwarf_expand_die (&cu->dies[cu->length]);
+		// reads all attribute valeus based on abbreviation
+		for (i = 0; i < curr_abbr->length - 1; i++) {
+			if (curr_die->length == curr_die->capacity) {
+				expand_die (curr_die);
 			}
-			if (i >= cu->dies[cu->length].capacity || i >= da->decls[abbr_code - 1].capacity) {
+			if (i >= curr_die->capacity || i >= curr_abbr->capacity) {
 				eprintf ("Warning: malformed dwarf attribute capacity doesn't match length\n");
 				break;
 			}
-			memset (&cu->dies[cu->length].attr_values[i], 0, sizeof (cu->dies[cu->length].attr_values[i]));
-			buf = r_bin_dwarf_parse_attr_value (buf, buf_end - buf,
-					&da->decls[abbr_code - 1].specs[i],
-					&cu->dies[cu->length].attr_values[i],
-					&cu->hdr, debug_str, debug_str_len);
-			if (cu->dies[cu->length].attr_values[i].name == DW_AT_comp_dir) {
-				const char *name = cu->dies[cu->length].attr_values[i].encoding.str_struct.string;
+			memset (&curr_die->attr_values[i], 0, sizeof (curr_die->attr_values[i]));
+
+			buf = parse_attr_value (buf, buf_end - buf,
+				&curr_abbr->defs[i],
+				&curr_die->attr_values[i],
+				&unit->hdr, debug_str, debug_str_len);
+
+			bool is_string_form = curr_die->attr_values[i].attr_form == DW_FORM_strp ||
+				curr_die->attr_values[i].attr_form == DW_FORM_string;
+
+			// TODO  does this have a purpose anymore?
+			// Or atleast it needs to rework becase there will be
+			// more comp units -> more comp dirs and only the last one will be kept
+			if (curr_die->attr_values[i].attr_name == DW_AT_comp_dir &&
+				is_string_form &&
+				curr_die->attr_values[i].string.content) {
+				const char *name = curr_die->attr_values[i].string.content;
 				if ((size_t)name > 1024) { // solve some null derefs
-					sdb_set (s, "DW_AT_comp_dir", name, 0);
+					sdb_set (sdb, "DW_AT_comp_dir", name, 0);
 				} else {
 					eprintf ("Invalid string pointer at %p\n", name);
 				}
 			}
-			cu->dies[cu->length].length++;
+			curr_die->length++;
 		}
-		cu->length++;
+		unit->length++;
 	}
 	return buf;
 }
 
-R_API int r_bin_dwarf_parse_info_raw(Sdb *s, RBinDwarfDebugAbbrev *da,
+/**
+ * @brief Reads all information about compilation unit header
+ * 
+ * @param buf Start of the buffer
+ * @param buf_end Upper bound of the buffer
+ * @param unit Unit to read information into
+ * @return ut8* Advanced position in a buffer
+ */
+static const ut8 *info_comp_unit_read_hdr(const ut8 *buf, const ut8 *buf_end, RBinDwarfCompUnitHdr *hdr) {
+	// 32-bit vs 64-bit dwarf formats
+	// http://www.dwarfstd.org/doc/Dwarf3.pdf section 7.4
+	hdr->length = READ32 (buf);
+	if (hdr->length == (ut32)DWARF_INIT_LEN_64) { // then its 64bit
+		hdr->length = READ64 (buf);
+		hdr->is_64bit = true;
+	}
+	const ut8 *tmp = buf; // to calculate header size
+	hdr->version = READ16 (buf);
+	if (hdr->version == 5) {
+		hdr->unit_type = READ8 (buf);
+
+		hdr->address_size = READ8 (buf);
+
+		if (hdr->is_64bit) {
+			hdr->abbrev_offset = READ64 (buf);
```

Comment:
I think a ternary operator is much more readable here.
```C
hdr->abbrev_offset = hdr->is_64bit? READ64(buf): READ32 (buf);
```

Reply:
Or, even better, since this is used in a bunch of places just use a separate function and use it for the other places as well.

Reply:
Ye, I should do{ } while() the macros there to be able to do that. As for the function, it's really hard to figure out how to name such function :D

Reply:
`dwarf_read_piece(bool is_wide)` or something like that.